# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
df_children_debt = data.groupby('children')['debt'].agg([
        ('num_clients', 'count'),    
        ('debt_count', 'sum'),       
        ('debt_rate', 'mean')        
    ])

df_children_debt.sort_values(by='debt_rate', ascending=False)

num_clients  debt_count  debt_rate
children                                    
4                  41           4   0.097561
2                2052         194   0.094542
1                4808         444   0.092346
3                 330          27   0.081818
0               14091        1063   0.075438
5                   9           0   0.000000

**Вывод:** \
Можем заметить, что с увеличением количества детей процент людей, что имели задолженности по возврату кредита, возрастает. Однако в то же время семьи, у кого 3 ребенка, чаще погашают задолженности вовремя, как и те, у кого детей совсем нет. _На основе имеющихся данных довольно трудно сделать однозначные выводы_.\
Возможно, необходимо несколько больше данных, например, по тем клиентам банка, у кого трое детей. Это помогло бы лучше подтвердить гипотезу об однозначной зависимости между количеством детей и возвратом кредита в срок

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
df_fs_debt = data.groupby('family_status')['debt'].agg([
        ('num_clients', 'count'),    
        ('debt_count', 'sum'),       
        ('debt_rate', 'mean')        
    ])

df_fs_debt.sort_values(by='debt_rate', ascending=False)

num_clients  debt_count  debt_rate
family_status                                            
Не женат / не замужем         2796         273   0.097639
гражданский брак              4134         385   0.093130
женат / замужем              12261         927   0.075606
в разводе                     1189          84   0.070648
вдовец / вдова                 951          63   0.066246

**Вывод:**\
Да, здесь зависимость есть. Люди, не состоявшие в браке, _реже возвращают кредит в срок_, чем те, кто на данный момент женат/замужем. Более того, можем заметить также то, что те люди, кто прежде состоял в браке, имеют лучшие показатели по возврату кредита по срокам

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
income_info = data['total_income'].agg(['min', 'max', 'mean', 'median'])
income_info.apply(lambda x: f'{x:,.2f}')

min          20,667.00
max       2,265,604.00
mean        165,342.75
median      142,594.00
Name: total_income, dtype: object

In [25]:
data['total_income_category'].value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

**Изучим показатели ежемесячных доходов клиентов** \
Мы имеем дело с сильно смещённым распределением с высокими выбросами (разрыв между медианой и максимумом довольно огромный), что на самом деле кажется типичным для доходов.
Для лучших выводов нам следует провести анализ по категориям «B» и «C», ввиду того, что данных по остальным категориям довольно мало (высокая статистическая погрешность)


In [26]:
df_income_debt = data.pivot_table(
    index='total_income_category',
    columns='debt',
    values='total_income',
    aggfunc='count'
)

df_income_debt.columns = ['без просрочки', 'с просрочкой']

df_income_debt['всего данных'] = (
    df_income_debt['без просрочки'] + df_income_debt['с просрочкой']
)

df_income_debt['Доля просроченных кредитов (%)'] = (
    df_income_debt['с просрочкой'] / df_income_debt['всего данных'] * 100
)

df_income_debt


без просрочки  с просрочкой  всего данных  \
total_income_category                                              
A                                 23             2            25   
B                               4660           354          5014   
C                              14568          1353         15921   
D                                328            21           349   
E                                 20             2            22   

                       Доля просроченных кредитов (%)  
total_income_category                                  
A                                            8.000000  
B                                            7.060231  
C                                            8.498210  
D                                            6.017192  
E                                            9.090909

**Выводы**

Разбивка клиентов по категориям дохода выглядит так:

* **E**: 0–30 000
* **D**: 30 001–50 000
* **C**: 50 001–200 000
* **B**: 200 001–1 000 000
* **A**: свыше 1 000 000

Доля просроченных кредитов варьируется в разных группах, и в некоторых категориях выборка слишком мала для достоверных выводов. Тем не менее в основных сегментах (B и C), которые составляют большинство клиентов, отчётливо заметна обратная зависимость: с ростом дохода процент должников снижается. _Чем выше доход клиента, тем меньше вероятность просрочки._


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [27]:
df_children_debt = data.groupby('purpose_category')['debt'].agg([
        ('num_clients', 'count'),    
        ('debt_count', 'sum'),       
        ('debt_rate', 'mean')        
    ])

df_children_debt.sort_values(by='debt_rate', ascending=False)

num_clients  debt_count  debt_rate
purpose_category                                            
операции с автомобилем           4279         400   0.093480
получение образования            3988         369   0.092528
проведение свадьбы               2313         183   0.079118
операции с недвижимостью        10751         780   0.072551

**Вывод:**\
Цели получения кредитов можно разбить на 4 категории:
 - операции с автомобилем
 - получение образования
 - проведение свадьбы
 - операции с недвижимостью
 
Ранее мы уже произвели группировку целей кредитов по этим категориям, поэтому можем довольно просто вывести необходимую информацию по возврату кредита в сроки для группы пользователей.

Здесь мы можем заметить, что цели, связанные с недвижимостью и свадьбой, имеют довольно низкий процент просроченных сроков кредитования, то есть клиенты этих категорий чаще возвращают деньги вовремя. Деньги же, взятые для трат на автомобиль и образование, возвращают вовремя реже.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* отметим колонки, где имелись пропущенные значения:
 - total_income
 - days_employed

Обратившись к блоку с предобработкой данных, можем заметить, что в нашем датасете пропуски в этих двух столбцах равны одному и тому же значению (2174). Это значит, что пропущенные значения не случайны, а, скорее всего, связаны с определёнными категориями клиентов или этапами сбора данных.
Я могу выделить следующие возможные причины:
1. Пенсионеры, студенты и безработные
 - У пенсионеров официальный доход зачастую не привязан к зарплате, а выплачивается как пенсия или субсидия, и банк мог хранить такой доход в других колонках (или вообще не считать его “рабочим” доходом).
 - У студентов и безработных фактический «общий трудовой стаж в днях» либо равен нулю, либо не имеет смысла (они просто не работали официально). Банк мог оставить эти строки пустыми, а не прописывать «0», чтобы не вводить в заблуждение аналитиков.
 - Возможно, безработные (те, кто не проработал ни одного дня) не отображаются как «0» в базе данных банка, а просто трактуется как отсутствие информации о клиенте. Посмотрим минимальное значение в столбце days_employed:

In [28]:
data['days_employed'].min() # минимальное значение общего трудового стажа в днях у клиентов

24.14163324048118

Это подтверждает предположение последнего пункта про сохранение данных о безработных.

2. Конфиденциальность и внутренние правила банка
- Некоторые клиенты могли дать согласие на получение кредита, но не дали разрешение банку хранить или обрабатывать их данные о заработке, ввиду чего пришлось оставить значения ежемесячного дохода и трудового стажа пустыми

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Это решение можно считать оптимальным по нескольким причинам:
1. Сильно скошенные данные (мы работаем с ежемесячными доходами, в значениях которых зачастую можно найти выбросы, а также с общим трудовым стажем пенсионеров и безработных, где тоже наблюдались аномальные значения, которые мы решили не исправлять).

2. Небольшое число пропусков относительно всего объёма (тогда замена одной константой не сильно исказит общую картину).

3. Это довольно быстрая предобработка данных до более сложного анализа. Особенно оказывается отличным решением в тот момент, когда нет много времени строить модели для предсказания отсутствующих значений.

### Шаг 4: общий вывод.

_В результате нашей работы мы проделали следующее:_
1. Заполнили пропуски по столбцам days_employed и total_income (медианными значениями по каждого типа занятости income_type), а также дали объяснения по возможному происхождению этих пропусков и по тому, ввиду чего стоит брать именно медианные значения для них
2. Обработали аномальные отрицательные значения в столбце days_employed, взяв их противоположные значения, а также аномальные данные в столбце children, удалив их
3. Обработали дубликаты, убедившись, что в столбце education нет повторяющихся значений по клиентам в разном регистре (Высшее и высшее - должно быть одним и тем же)
4. Произвели категоризацию данных:
 - Разбили доход клиентов на 5 категорий
 - На основании данных по клиентам столбца purpose (цель кредитования) составили 4 основные категории получения кредита.


_На основе полученной информации мы можем заключить следующее о выявленных зависимостях просрочки от разных факторов:_
1. **Семейное положение**
    1. Надежные заемщики:
          1. Клиенты, состоявшие в браке ранее (разведенные/овдовевшие), демонстрируют наименьшую долю просрочек.
          2. Те, кто находится в браке (женат/замужем), также платят вовремя чаще, чем незамужние/неженатые.
    2. Рискованные заемщики:
          1. Клиенты, никогда не состоявшие в браке, имеют более высокую вероятность просрочки по кредиту.
2. **Количество детей**
    1. Надежные заемщики:
          1. Клиенты без детей 
    2. Рискованные заемщики:
          1. Клиенты с 1 или 2 детьми - их доля просроченных кредитов несколько выше, нежели чем у семей без детей. 
          2. По клиентам с 3 и более детьми выборка недостаточна для однозначных выводов.
3. **Цель кредита**
    1. Надежные цели:
        1. Покупка недвижимости и проведение свадьбы — наименьший риск просрочки.

    1. Рискованные цели:
        1. Приобретение автомобиля и дополнительное образование — более высокая доля просрочек.
4. **Доходы**\
    В группах с достаточным объёмом данных (категории B и C) видно, что клиенты из категории B (200 001–1 000 000 ₽) имеют более низкий процент просрочек (≈ 7,06 %) по сравнению с категорией C (50 001–200 000 ₽) (≈ 8,50 %). Это говорит о том, что в средне–высоком сегменте дохода (категория B) доля должников ниже, чем в среднем сегменте (категория C). В крайних категориях (E и A) выборки слишком малы, чтобы делать статистически надёжные обобщения, а категория D (30 001–50 000 ₽) показывает долю ≈ 6,02 %, однако её объём тоже сравнительно небольшой, чем в B и C, поэтому она менее репрезентативна. В таком случае можем сделать следующее заключение:
    - Между группами B и C, где статистика наиболее надёжна, наблюдается: чем выше доход (переход из C в B), тем ниже процент просрочек.
    - В других категориях (E, D, A) выборка весьма мала, чтобы можно было сделать надежные выводы
    
На основе проведенного анализа мы можем выделить ключевые характеристики, которые отличают **надежных** заемщиков от **рискованных**. Эти выводы помогут банку оптимизировать кредитную политику, минимизировать риски и принимать более обоснованные решения.

_Портреты заемщиков_
1. Надежный заемщик:

    1. Семейное положение: вдовец/разведенный или официально женат/замужем.

    2. Дети: детей нет

    3. Цель кредита: операции с недвижимостью, свадьба.
    
    4. Доход: преимущественно в категории B (200 001–1 000 000 ₽).


2. Рискованный заемщик:

    1. Семейное положение: никогда не состоял в браке.

    2. Цель кредита: покупка автомобиля, образование.
    
    3. Доход: чаще встречается в средних (<200 000 ₽) или низких (<50 000 ₽) категориях, но надёжность дохода менее критична, чем плохие сочетания семейного статуса и цели.
    
Таким образом, стоит активнее работать с семейными клиентами (особенно вдовцами/разведенными) и теми, кто берет кредиты на недвижимость или свадьбу — предлагать им выгодные условия. Для займов на авто, образование и многодетных семей — усилить контроль и ограничить риски, так как их доля просрочек выше. Кроме того, при оценке потенциальных должников учитывать доход преимущественно в категориях B и C, понимая, что в сегменте B (200 001–1 000 000 ₽) клиент надёжнее, чем в C (50 001–200 000 ₽).

_Данные показывают: ключевые факторы при работе с клиентами — семейный статус, цели, количество детей и доходность клиента._
